### ISSUE TITLE
load-into-countgraph rename messes up multiple versions.

### ISSUE THREAD

**Issue by ctb** [dib-lab/khmer#1256]
If you have an older version of khmer, and you install 2.0-pre, load-into-counting runs the old version and load-into-countgraph runs the new version.

More generally, we need to discuss script *renaming* in a top level issue.  All of our documentation and examples - much of it not linked into khmer directly - depends on these script names.

I'm -1 on changing script names at this point in the release cycle.

If there are any other such renames, please mention them here.

**Comment by ctb** [comment]
The underlying logic behind the -1 is user focused, BTW - changing command line options is frustrating for users but necessary sometimes (and a major version # definitely is a good reason).  However, if we change the script name users don't know what script to look for, and they may accidentally run older versions that still import the newer package code.  There are probably other undesirable outcomes.  So I think it's a bigger decision than "hey, I'd like to rename this script to be consistent with our internal nomenclature."

**Comment by ctb** [comment]
Same with load-into-nodegraph rename. Commits 2d621f6 and fa75f9cb342.

**Comment by ctb** [comment]
Here are the commands to reproduce the bug that alerted me to this; run these on a blank Ubuntu 14.04
```
sudo apt-get update && \
sudo apt-get -y install screen git curl gcc make g++ python-dev unzip \
           default-jre pkg-config libncurses5-dev r-base-core \
           r-cran-gplots python-matplotlib sysstat python-virtualenv \
           python-setuptools cmake

cd
python -m virtualenv env
source env/bin/activate
pip install -U setuptools
pip install khmer==1.4.1

# at some later point, decide to update:
git clone https://github.com/dib-lab/khmer.git
cd khmer
make install

cd /mnt
load-into-counting.py ...
```

This last command fails with:
```
Traceback (most recent call last):
  File "/home/ubuntu/env/bin/load-into-counting.py", line 26, in <module>
    from khmer.kfile import check_space_for_hashtable
ImportError: cannot import name check_space_for_hashtable
```

This seems broken to me - I didn't realize that pip didn't lock things to version (i.e. I would have expected load-into-counting.py to load the khmer library from 1.4.1).  But it also adds a reason to not rename scripts, I think.

An alternative to renaming scripts is to produce scripts with multiple names, e.g. load-into-counting and load-into-countgraph, OR have a replacement 'load-into-counting' that says "use load-into-countgraph" - 

I'm -0 on changing the name at all, I'm -1 on eliminating load-into-counting.

**Comment by ctb** [comment]
(This is a blocker for 2.0.)

**Comment by mr-c** [comment]
Changing script names will happen eventually so I'm first going to address that issue.

This is a setuptools and not a pip bug. If one upgrades using pip it correctly uninstalls the older version first. We can add a `pip uninstall` loop to the beginning of the `make install` target to emulate the correct behavior.

```
(sid) mcrusoe@localhost:~/khmer$ virtualenv testenv
New python executable in testenv/bin/python
Installing setuptools, pip...done.
(sid)mcrusoe@localhost:~/khmer$ source testenv/bin/activate
(testenv)(sid)1 mcrusoe@localhost:~/khmer$ pip install khmer
Collecting khmer
  Using cached khmer-1.4.1.tar.gz
Requirement already satisfied (use --upgrade to upgrade): screed==0.8 in ./testenv/lib/python2.7/site-packages (from khmer)
Requirement already satisfied (use --upgrade to upgrade): bz2file in ./testenv/lib/python2.7/site-packages (from screed==0.8->khmer)
Installing collected packages: khmer
  Running setup.py install for khmer
Successfully installed khmer-1.4.1
(testenv)(sid)mcrusoe@localhost:~/khmer$ pip install dist/khmer-2.0rc3+66.g94d679f-cp27-none-linux_x86_64.whl 
Processing ./dist/khmer-2.0rc3+66.g94d679f-cp27-none-linux_x86_64.whl
Requirement already satisfied (use --upgrade to upgrade): bz2file in ./testenv/lib/python2.7/site-packages (from khmer==2.0rc3+66.g94d679f)
Collecting screed>=0.9 (from khmer==2.0rc3+66.g94d679f)
Installing collected packages: screed, khmer
  Found existing installation: screed 0.8
    Uninstalling screed-0.8:
      Successfully uninstalled screed-0.8
  Found existing installation: khmer 1.4.1
    Uninstalling khmer-1.4.1:
      Successfully uninstalled khmer-1.4.1
Successfully installed khmer-2.0rc3+66.g94d679f screed-0.9
(testenv)(sid)mcrusoe@localhost:~/khmer$ find testenv/ | grep load-
testenv/bin/load-into-countgraph.py
testenv/bin/load-into-nodegraph.pyc
testenv/bin/load-into-nodegraph.py
testenv/bin/load-into-countgraph.pyc
```

**Comment by ctb** [comment]
On Mon, Aug 17, 2015 at 06:51:22PM -0700, Michael R. Crusoe wrote:
> Changing script names will happen eventually so I'm first going to address that issue.

In the preferred future, we'll be replacing this with the oxli script
and subcommands (#389), which would be accompanied by a renaming of the
project.

**Comment by mr-c** [comment]
Using a `diff`d output of `--help` against all the scripts in both the v1.4.1 and the current (c7c7606fd7c5d2fa92524028a2e3fbcf93de1caf) vesion, here are the backwards incompatible changes we have committed so far with respect to the command line interface:

`--min-tablesize` has been replaced with `--max-tablesize` in `abundance-dist-single.py`, `do-partition.py`, `filter-abund-single.py`, `find-knots.py`

`--savetable` is now `--savegraph` in `abundance-dist-single.py`, `filter-abund-single.py`

`--report-total-kmers` option eliminated from `abundance-dist-single.py`, `filter-abund-single.py`

`--csv` option eliminated from `abundance-dist-single.py`

The script `count-overlap.py` has been eliminated.

`interleave-reads.py

**Comment by mr-c** [comment]
Using a `diff`d output of `--help` against all the scripts in both the v1.4.1 and the current (c7c7606fd7c5d2fa92524028a2e3fbcf93de1caf) vesion, here are the backwards incompatible changes we have committed so far with respect to the command line interface:

`--min-tablesize` has been replaced with `--max-tablesize` in `abundance-dist-single.py`, `do-partition.py`, `filter-abund-single.py`, `find-knots.py`, `load-into-[countgraph|counting].py`, `load-[into-node]graph.py`, `make-initial-stoptags.py`, `normalize-by-median.py`, `trim-low-abund.py`

`--savetable` is now `--savegraph` in `abundance-dist-single.py`, `filter-abund-single.py`, `normalize-by-median.py`, `trim-low-abund.py`

`--report-total-kmers` option eliminated from `abundance-dist-single.py`, `filter-abund-single.py`, `normalize-by-median.py`

`--csv` option eliminated from `abundance-dist-single.py`

The script `count-overlap.py` has been eliminated.

`interleave-reads.py` will no longer auto-guess the name of the second file if only one is given.

`--write-fp-rate` has been eliminated as an option from `load-into-[countgraph|counting].py`

`--save-on-failure` & `--dump-frequency` have been eliminated from `normalize-by-median.py`

`--out` is now `--output` for `normalize-by-median.py`, `trim-low-abund.py`

`--loadtable` is now `--loadgraph` in `normalize-by-median.py`, `trim-low-abund.py`

`-p`/`--force-paired` has been eliminated from `split-paired-reads.py`

**Comment by mr-c** [comment]
`table`->`graph` renames were my idea, spurned on by the general rename going on in the code and the fact that so many other API breaking changes had occurred. 

Ignoring those changes still leave 20 other API breaking changes between v1.4.1 and the current development version.

**Comment by mr-c** [comment]
I have already shown that users won't be impacted by the script renaming. We can send a note out alerting developers that they may be better served with a fresh virtualenv.

**Comment by mr-c** [comment]
I'm happy to add pass-through or `FYI` shims under the old names. This would likely also solve the problem some developers might run into.

**Comment by mr-c** [comment]
A PR to update the `whats-new-2.0.rst` docs with my findings above is forthcoming.

**Comment by ctb** [comment]
But the rename is still just pointless; if we were sticking with the scripts for a few more big versions, I might see a point, but causing user pain and massive documentation breakage solely for some fairly abstract notion of consistency that will massively change again within a year doesn't strike me as a worthwhile idea. It's tripped up two experienced users, as well, which points in a very different direction than the blithe assumption you are making that it's a minor change.  It also was done at your whim and without any real discussion or evaluation of the pros and cons, and that is not the way we should proceed. And, finally, it was done in the last minute before a new major version, and I'm not ok with that kind of process.

So, please revert it and punt it to an issue if you want us to discuss it as a project. I would prefer to talk about the oxli interface myself.

--
Titus Brown, ctbrown@ucdavis.edu

**Comment by mr-c** [comment]
I am sorry to hear that you think it was pointless, that I was making blythe assumptions about the competency of our users, and that I did this on a whim. I do agree that it is worthy of discussion which is what we are doing now.

I also agree that the oxli interface is more interesting but is outside of my scope at the moment.

To unblock the 2.0 release I've made a PR to undo this change. #1270. Your assistance is appreciated.
